### ISSUE TITLE
Report attributes with joins

### ISSUE THREAD

**Issue by pithyless** [fulcrologic/fulcro-rad#28]
I'm struggling with using RAD Reports in anger and it mostly has to do with limitations around how the attributes are used as relations. Before the API around `defsc-report` is finalized, I'd like to add my 2 cents about a change to the exisiting reports information model. If this is an avenue worth pursuing, I'd be happy to help with submitting PRs to move us closer to the goal, but first I'd like to flush out if this idea makes sense.

I felt this was too long of a rambling for a Slack thread, so I'm submitting this as an issue to discuss.

# Status Quo

```clojure
;; Fig-1
(report/defsc-report InventoryReport [this props]
  {ro/source-attribute    :item/all-items
   ro/row-pk              item/id
   ro/columns             [item/item-name
                           category/label]})
```

This implies the following EQL query that will be processed by the server:

```clojure
;; Fig-2
[{:item/all-items
  [:item/id :item/item-name :category/label]}]
```

In practice, a Pathom parser can be implemented this way:

```clojure
;; Fig-3
[{:pc/output [{:item/all-items
               [:item/id :item/item-name :category/label]}]}]
```

Or this way:

```clojure
;; Fig-4
[{:pc/input  #{:item/id}
  :pc/output [:item/item-name :category/id]}
 {:pc/input  #{:category/id}
  :pc/output [:category/label]}
 {:pc/output [{:item/all-items [:item/id]}]}]
```

I am highly in favor of Fig-4 over Fig-3, as a way of reducing boilerplate code,
centralizing authorization code, etc.

# Flattening Data Fetching

The approach to reports in Sec-1 works, because it flattens the data model to remove a level of nesting in to-one relations (category has many items, but item has only one easily understood category).

Sometimes these relationships are not so obvious:

```clojure
;; Fig-5
[{:all-items
  [:item/name
   :employee/name   ;; an employee signs off the purchase
   :company/name    ;; a company packages items
   :company/name    ;; another company ships it???
   ]}]
```

The relationship between employee and item is not immediately obvious (as it was with category), and using two companies (for different semantic purposes) is currently not possible.  EQL is wonderfully flexible, and we can solve this either via flattened namespaces:

```clojure
;; Fig-6
[{:all-items
  [:item/name
   :item.signed-off-by/name ;; an employee signs off the purchase
   :item.packaged-by/name   ;; a company packages items
   :item.shipped-by/name    ;; another company ships it
   ]}]
```

OR via joins:

```clojure
;; Fig-7
[{:all-items
  [:item/name
   {:item/signed-off-by [:employee/name]} ;; an employee signs off the purchase
   {:item/packaged-by [:company/name]}    ;; a company packages items
   {:item/shipped-by [:company/name]}      ;; another company ships it
   ]}]
```

I'm especially interested in opinions whether we should be modeling our APIs more like Fig-6 or Fig-7. Fig-6 seems closer to what @wilkerlucio  is proposing in talks like [The Maximal Graph](https://www.youtube.com/watch?v=IS3i3DTUnAI), but Fig-7 has the advantage that we can reuse a lot of the existing Pathom plumbing on the server and Fulcro DB caching on the client-side.

If we were to go down the road of Fig-6, we need to create custom aliases for resolvers and mutations for every kind of relation we want to see in reports and forms. If we choose to follow Fig-7, we need some way of declaring joins. In either case, we need some kind of extension to the existing `defsc-report`.

# Loose Proposal

Assuming we would like to explore supporting Fig-6 or Fig-7 use-cases, we may consider separating virtual report-columns (what we see), from actual attributes (what we fetch). One suggestion may look something like this:

```clojure
;; Fig-8
(report/defsc-report InventoryReport [this props]
  {ro/source-attribute    :item/all-items
   ro/row-pk              item/id
   ro/columns             [{:column :name
                            :attr   item/item-name}
                           {:column :label
                            :attr   category/label
                            :join   [:item/category]}]})
```

By separating the identity of the column in the report from the data-source, we
can support more kinds of data columns (see Fig-6 or Fig-7). This would also
give us a consistent place to extend the config, eg. optional column headers:

```clojure
Fig-9
(report/defsc-report InventoryReport [this props]
  {ro/source-attribute    :item/all-items
   ro/row-pk              item/id
   ro/columns             [{:column  :name
                            :attr    item/item-name
                            :title   "Item Name"}
                           {:column  :label
                            :attr    category/label
                            :join    [:item/category]
                            :title   "Category"}]})
```

And the macro can still support a terse version of the column syntax, for
backwards compatibility and when it's good enough:

```clojure
Fig-10
(translate-columns
  {:ro/columns [item/item-name]})
;; =>
{:ro/columns [{:column :item/item-name
               :attr   item/item-name
               ...}]}
```

# TL;DR

Does this make sense? Is this worth pursuing? It would be a shame to have to write custom components to support this kind of data-fetching and lose all the capabilities that the reports already provide. If it's just a question of lack of time, I'll be happy to chip in and do the grunt work if we can just agree on the direction.

**Comment by awkay** [comment]
I appreciated you taking the time to write all of this, and I appreciate that you took the time to understand the problems we're facing the solving.

For *certain* kinds of reports (really just lists of hard entity facts) the normalization aspect would be nice. For derived data reports it is simple going to be resolvers.  I also agree with you that while the *current* approach is to use flattened space with new keywords, it would be nice to have some concise way to describing the path to data that we're interested in so that we could leverage normalization fully.

Remember, however, that normalization requires generation of mock components for each one of those generated joins, because that is how you supply normalization with its data. It is simple enough to do, but a bit of a hassle to manage internally.

So, the real question is on API.

At the moment, the parser solution is quite good enough for most purposes, and I want to stay with columns *as-is*. There is no need to complicate the common use case.  

However, being able to designate paths might be as simple as allowing nested vectors of attributes:

```
::report/columns [invoice/date [invoice/packaged-by company/name] invoice/total]
```

where the elements in the sub-vector must all be of `:ref` `:one` type except the last.

The existing column-headers attribute could be similarly extended to allow a vector of keywords to designate the late...same with column formatters.

Thoughts?
```
Then, an additional attribute can find some way to supply the *path* to get those attributes, which is extra information that is optional and only needed if we know our back-end parser will not be able to figure things out:

```
::report/data-paths {:

**Comment by pithyless** [comment]
Thanks @awkay for the fast and thorough response.

The nested vector attributes is not an approach I had considered, and it would (possibly?) imply a lot of internal code that needs to be checking if we're looking at an attribute or a seq of attributes. But, that's just my first gut response. I'm willing to look past it, if it's more aligned with your own API aesthetics. :)

On first glance, I can't think of a counter-example why this approach wouldn't work, per se. And forcing all normalization joins to be first-class attributes may be a better overall design.

Are you currently working on refactoring the reports further? Or interested in pushing this forward? If it's a low priority for you, I can try to take a stab at making a prototype.

**Comment by awkay** [comment]
I would love the help. When are you available to chat? PM me in Slack.

**Comment by awkay** [comment]
Closed due to inactivity.
